doc/developer/design/20231127_decoupled_isolated_coordinator.md

```diff
@@ -0,0 +1,472 @@
+  AppendError>`: appends the given updates to the catalog _iff_ the expected
+  upper matches its current upper.
+
+- `SubscribeAt(timestamp) -> Subscribe`: returns a snapshot of the catalog at
+  `timestamp`, followed by an ongoing stream of subsequent updates.
+
+- `SnapshotAt(timestamp)`: returns an in-memory snapshot of the catalog at
+  `timestamp`.
+
+  A snapshot corresponds to our current in-memory representation of the catalog
+  that the Coordinator holds. This can be (and will be, in practice!)
+  implemented using `SubscribeAt` to incrementally keep an in-memory
+  representation of the catalog up to date with changes. However, it is
+  logically a different operation.
+
+### The controllers, aka. STORAGE and COMPUTE
+
+The controllers remain largely as they have been previously designed, but
+query-like methods gain a new `command_as_of` parameter and we introduce a new
+`DowngradeCommandCapability` method. The controllers can be seen as consuming a
+stream of commands and `DowngradeCommandCapability` is downgrading its
+frontier. The `command_as_of` parameter on query-like operations tells the
+controller that it can only process them once its input frontier has
+sufficiently advanced.
+
+These changes make it so that the controller knows up to which CATALOG
+timestamp it has received commands and to ensure that peeks (or other
+query-like operations) are only served once we know that controller state is up
+to date with the catalog state as of which the peek was processed.
+
+As we see below in the section on ADAPTER. We replace explicit sequencing by
+decoupling using pTVCs and timestamps.
+
+- `Peek(global_id, command_as_of, ..)`: performs the peek, but _only_ once the
+  command capability has been downgraded far enough.
+
+  It is important to note that the new `command_as_of` parameter is different
+  from an `as_of`, which latter refers to the timeline/timestamp of the object
+  being queried.
+
+- `DowngradeCommandCapability(timestamp): tells the controller that it has seen
+  commands corresponding to catalog changes up to `timestamp`.
+
+(Most existing commands have been omitted for brevity!)
+
+### ADAPTER
+
+ADAPTER spawns threads/processes/tasks for handling client connections as they
+show up.
+
+Additionally, it spawns one task per controller that runs a loop that
+synthesizes commands for the controller based on differential CATALOG changes.
+To do this, the task subscribes to CATALOG, filters on changes that are
+relevant for the given controller, sends commands to the controller while
+preserving the order in which they are observed in the stream of changes, and
+synthesizes `DowngradeCommandCapability` commands when the catalog timestamp
+advances. These commands are the commands that deterministically derive from
+CATALOG state (see Background section, above).
+
+When processing client requests, ADAPTER uses TIMESTAMP ORACLE to get the
+latest read timestamp for CATALOG, then fetches a CATALOG snapshot as of at
```

Comment:
> then fetches a CATALOG snapshot as of at least that timestamp

This makes it sound like the catalog's upper has to be advanced continually as time passes, and not just when the catalog changes. Is that right? If so, we might want to address why we're not making the catalog upper lazy in the alternatives section (I'm certainly curious)

Reply:
I've been assuming it would advance only when there are changes, and only when there are changes would the write timtestamp (at the oracel) be bumped. So most of the time you would get the same read timestamp for the catalog

Reply:
Okay cool. This bit I quotes sound a bit like it contradicts that, so probably worth massaging the language here a bit